mesonbuild/interpreter.py

```diff
@@ -3126,7 +3126,7 @@ def func_environment(self, node, args, kwargs):
     @stringArgs
     @noKwargs
     def func_join_paths(self, node, args, kwargs):
-        return os.path.join(*args).replace('\\', '/')
```

Comment:
@nirbheek how does this tie in with #3400?

Reply:
@jpakkane I need the paths to be consistent so that I can do either os.path.split or str.split('/'). I found this fix easier than removing the 549 occurrences of os.path.join().

```
$ rg os.path.join mesonbuild/ | wc -l
549
```

If only `/` is supposed to be used as a separator then IMHO you cannot use os.path.join anywhere as it will introduce `\\` on windows.

Reply:
This change will probably cause breakage.

Instead of using `os.path.split()`, you should use `pathlib.Purepath().parts`. On Windows that will split on both `/` and `\` (even if you mix them in the same path):

```python
>>> from pathlib import PurePath
>>> p = PurePath('C:/Windows/System32')
>>> p
PureWindowsPath('C:/Windows/System32')
>>> p.parts
('C:\\', 'Windows', 'System32')
>>> q = PurePath('C:\\Windows\\System32')
>>> q
PureWindowsPath('C:/Windows/System32')
>>> q.parts
('C:\\', 'Windows', 'System32')
```

We should use `/` for all paths everywhere, but as you say there's a lot of `os.path.join()` everywhere that we should slowly change. Not something we should do this late.

Reply:
Note that to convert a `PurePath()` back to a string, you should always use `.as_posix()` otherwise on Windows it will return a string with `\\` as the separator.

Reply:
It doesn't break any tests but maybe the tests aren't exercising join_paths much.

Reply:
Using PurePath doesn't solve anything:

```
>>> import pathlib
>>> pathlib.PurePath('C:\\Program Files/test')
PurePosixPath('C:\\Program Files/test')
>>> p = pathlib.PurePath('C:\\Program Files/test')
>>> p.as_posix()
'C:\\Program Files/test'
>>> help(p)
>>> p.parts
('C:\\Program Files', 'test')
>>> p = pathlib.PurePath('C:/Program Files\\test')
>>> p.parts
('C:', 'Program Files\\test')
>>> p.as_posix()
'C:/Program Files\\test'
```

Or do you mean that I should remove every single path with mixed directory separators (essentially `os.path.join()`) in the whole project?

Reply:
What Python are you testing with? Is this an MSYS2-provided Python or a Native-Windows Python?

Reply:
With `mingw32/mingw-w64-i686-python3` and `mingw64/mingw-w64-x86_64-python3`:

```python
>>> p = pathlib.PurePath('C:\\Windows/User')
>>> p.parts
('C:\\', 'Windows', 'User')
>>> p.as_posix()
'C:/Windows/User'
```

With `msys/python` (which is configured as UNIX-like, and hence won't understand `\`):

```python
>>> p = pathlib.PurePath('C:\\Windows/User')
>>> p.parts
('C:\\Windows', 'User')
>>> p.as_posix()
'C:\\Windows/User'
```

So I guess we should document somewhere that people should *never* use `msys/python` for anything.